### PR TITLE
Clean up IFI support and add library prereqs to modulefiles

### LINKED ISSUE #1057: IFI.F NetCDF diagnostic output takes lots of memory and runs slowly
The IFI.F can output NetCDF files in a format expected by libIFI's internal test programs. Presently, it works by gathering all variables on the first rank, and then writing them. Instead, it should do one variable at a time (ideally in a parallel write).

### LINKED ISSUE #1058: Compiling with libIFI should not require manually editing modulefiles
The external IFI build requires manually editing the modulefiles. This should be automated so UPP uses the right libIFI version.

### LINKED ISSUE #1064: Incorrect character(len=20) in read_netcdf_3d_para
The read_netcdf_3d_para in INITPOST_NETCDF.f uses a length of 20 for its "varname" argument. Callers send strings of length 4. This can lead to out-of-bounds reads. Also, it is flagged by the Intel compiler's debug mode.

### PR DESCRIPTION
**Pull request by SamuelTrahanNOAA**
This is part of an ongoing project to beautify and document the libIFI and associated programs so other people can develop the code. Changes are:

1. The IFI.F's NetCDF diagnostic file writer takes less memory and runs faster.
2. `compile_upp.sh -i` loads a modulefile for libIFI on Hera through `modulefiles/hera_external_ifi.lua`
3. Running `compile_upp.sh -I -B` will compile internal IFI and its internal test programs. This also loads the `hera_ifi_test_prereqs.lua` modulefile. That modulefile adds the NetCDF C++ wrapper, needed by libIFI's "fip2-lookalike.x" program.
4. Build system knows how to build the IFI executables, such as its test programs and the fip2-lookalike.x.
5. Updated the libIFI hash. (This is necessary for items 3 and 4 above.)
6. Improved logging and general code clean-up.

ALSO, a bug fix I accidentally left in these changes corrects bug #1064, an incorrect array length in INITPOST_NETCDF.f

Dependency on closed-source library PR. (Most users cannot access this link.)

- https://github.com/NCAR/UPP_IFI/pull/10

Issues:

- Fixes #1057 
- Fixes #1058
- Fixes #1064

### CODE REVIEW

File: tests/compile_upp.sh
Comment (WenMeng-NOAA) [inline tests/compile_upp.sh]: @SamuelTrahanNOAA It looks this line is duplicate of line 147.

**Comment by WenMeng-NOAA** [comment]
@SamuelTrahanNOAA The pre-installed libIFI is on Hera only?

**Comment by SamuelTrahanNOAA** [comment]
I'm installing the latest library on Jet now. Once Hera comes back, I'll update it there too.

My test data isn't on Jet anymore, so I have no way to test it at the moment. I'd have to copy the data and scripts back from Hera, which will take a few hours.

File: tests/compile_upp.sh
Comment (SamuelTrahanNOAA) [reply tests/compile_upp.sh]: I removed it.

**Comment by SamuelTrahanNOAA** [comment]
I don't own the libIFI installation on Acorn. The NCEP library group will need to update it there.

**Comment by SamuelTrahanNOAA** [comment]
The libIFI is on Jet, and I can confirm UPP on Jet can compile:

1. UPP with external IFI (compile_upp.sh -i)
2. UPP with internal IFI (compile_upp.sh -I)
3. The libIFI's own test programs (compile_upp.sh -I -B)

**Comment by WenMeng-NOAA** [comment]
@SamuelTrahanNOAA The build tests on Hera look good to me. Can you provide me the instructions of testing the executable ' fip2-lookalike.x' on Hera?

**Comment by SamuelTrahanNOAA** [comment]
> @SamuelTrahanNOAA The build tests on Hera look good to me. Can you provide me the instructions of testing the executable ' fip2-lookalike.x' on Hera?

There are two scripts.

* Revised script to run UPP and generate diagnostic files for input to fip2-lookalike.x
      /scratch2/BMC/ifi/Samuel.Trahan/fip/regtest/UPP/submit_run_post_rrfs_ifi_internal_hera.sh
* New script to test fip2-lookalike.x in the same directory where the post ran. It uses the diagnostic files as input.
      /scratch2/BMC/ifi/Samuel.Trahan/fip/regtest/UPP/submit_run_ifi_standalone_hera.sh

To confirm fip2-lookalike.x is working, you must compare:
* UPP output: cat_vars_0.nc
* fip2-lookalike output: icing-category-output.nc

The z0 variable is missing from one file, but the data within should be otherwise identical.

I'll update the code to output the missing z0 variable in a later PR.

**Comment by WenMeng-NOAA** [comment]
@SamuelTrahanNOAA I don't have access permission to /scratch2/BMC/ifi/Samuel.Trahan/fip.

**Comment by SamuelTrahanNOAA** [comment]
I copied them here. Can you see these?

* UPP internal libIFI test: /scratch2/BMC/ifi/Samuel.Trahan/for-wen/submit_run_post_rrfs_ifi_internal_hera.sh
* fip2-lookalike.x test: /scratch2/BMC/ifi/Samuel.Trahan/for-wen/submit_run_ifi_standalone_hera.sh

**Comment by WenMeng-NOAA** [comment]
@gspetro-NOAA @FernandoAndrade-NOAA You may start the UPP RTs on R&D machines. There should be no baseline changes with this PR.

**Comment by gspetro-NOAA** [comment]
@SamuelTrahanNOAA I can't push the RT logs to your branch. Could you give me permissions to do that?

**Comment by SamuelTrahanNOAA** [comment]
@gspetro-NOAA - I invited you as a collaborator to my repository. If you accept, you should be able to push logs.

@WenMeng-NOAA  - The libIFI pull request has no reviews. It needs to be merged first so I can point to a hash of the main branch instead of my feature branch.

**Review by gspetro-NOAA — COMMENTED** [review]
@WenMeng-NOAA All RTs pass on Hera, Orion, and Hercules with no baseline changes. Sounds like merging should wait until @SamuelTrahanNOAA 's dependent PR gets merged, so I'll hold off on hitting the "Approve" button, but once his PR gets merged, we should be good to go for this PR.

File: sorc/ncep_post.fd/INITPOST_NETCDF.f
Comment (WenMeng-NOAA) [inline sorc/ncep_post.fd/INITPOST_NETCDF.f:4053]: @SamuelTrahanNOAA Please add a change log with your name at the beginning session of this subroutine.

Comment (SamuelTrahanNOAA) [reply sorc/ncep_post.fd/INITPOST_NETCDF.f:4053]: The subroutine has no author information. Instead, all author information is at the top of the file. I added it there instead. Is that okay?

Comment (WenMeng-NOAA) [reply sorc/ncep_post.fd/INITPOST_NETCDF.f:4053]: @SamuelTrahanNOAA Yes, thanks for update.

Comment (SamuelTrahanNOAA) [reply sorc/ncep_post.fd/INITPOST_NETCDF.f:4053]: I left this change in by accident, but it is a bug fix. The character strings sent in are of length 4, not 20, which can produce an out-of-bounds read.

**Comment by SamuelTrahanNOAA** [comment]
I've made an issue for the bug fix in INITPOST_NETCDF.f where I changed len=20 to len=*

- #1064 

I left the bug fix in this branch by accident, but we may as well fix it here, rather than opening yet another PR that needs to be tested.

**Comment by SamuelTrahanNOAA** [comment]
@WenMeng-NOAA @gspetro-NOAA - The github GCC Linux Build test is failing with this error:

```
-- The Fortran compiler identification is unknown
/home/runner/work/UPP/UPP/spack/var/spack/environments/upp-env/.spack-env/view/bin/mpif90: line 386: /usr/bin/gfortran-10: No such file or directory
-- The C compiler identification is unknown
CMake Error at /usr/local/share/cmake-3.30/Modules/CMakeDetermineCXXCompiler.cmake:48 (message):
  Could not find compiler set in environment variable CXX:

  g++-10.

Call Stack (most recent call first):
  CMakeLists.txt:7 (project)


CMake Error: CMAKE_CXX_COMPILER not set, after EnableLanguage
-- Configuring incomplete, errors occurred!
```

**Comment by WenMeng-NOAA** [comment]
> @WenMeng-NOAA @gspetro-NOAA - The github GCC Linux Build test is failing with this error:
> 
> ```
> -- The Fortran compiler identification is unknown
> /home/runner/work/UPP/UPP/spack/var/spack/environments/upp-env/.spack-env/view/bin/mpif90: line 386: /usr/bin/gfortran-10: No such file or directory
> -- The C compiler identification is unknown
> CMake Error at /usr/local/share/cmake-3.30/Modules/CMakeDetermineCXXCompiler.cmake:48 (message):
>   Could not find compiler set in environment variable CXX:
> 
>   g++-10.
> 
> Call Stack (most recent call first):
>   CMakeLists.txt:7 (project)
> 
> 
> CMake Error: CMAKE_CXX_COMPILER not set, after EnableLanguage
> -- Configuring incomplete, errors occurred!
> ```

@AlexanderRichert-NOAA Do you have any suggestions to solve this GCC CI failure? Thanks!

**Comment by AlexanderRichert-NOAA** [comment]
Yes-- the ubuntu-latest runner now only has gcc 12, 13, and 14 available on accounting of now using ubuntu 24.04, so you can either update gcc to one of those versions, or revert to an earlier ubuntu version for the runner (I would suggest the former unless you specifically need gcc 10).

**Comment by SamuelTrahanNOAA** [comment]
You should choose a specific ubuntu version so our tests are consistent.

**Comment by SamuelTrahanNOAA** [comment]
This PR's dependency, in the UPP_IFI repository, has been merged.

**Comment by SamuelTrahanNOAA** [comment]
I'd prefer that a different PR update the github tests, not my PR.

Instead, I've changed the gnu.yml to explicitly specify the `ubuntu-22.04` image. That way, it tests exactly what it did before ubuntu-latest was changed to ubuntu-24.04.

In another PR, someone should update to a specific image version (ie. ubuntu-24.04) not a "latest" alias (ie. ubuntu-latest). Another linux distribution is fine. Also, I suggest you use a GCC version that matches what is used on one of the supported machines (hercules, hera, etc.)

**Comment by WenMeng-NOAA** [comment]
> This PR's dependency, in the UPP_IFI repository, has been merged.

@SamuelTrahanNOAA Have you updated the hash of libIFI.fd in this PR?

**Comment by SamuelTrahanNOAA** [comment]
> Have you updated the hash of libIFI.fd in this PR?

Now I have.

**Comment by WenMeng-NOAA** [comment]
> I'd prefer that a different PR update the github tests, not my PR.
> 
> Instead, I've changed the gnu.yml to explicitly specify the `ubuntu-22.04` image. That way, it tests exactly what it did before ubuntu-latest was changed to ubuntu-24.04.
> 
> In another PR, someone should update to a specific image version (ie. ubuntu-24.04) not a "latest" alias (ie. ubuntu-latest). Another linux distribution is fine. Also, I suggest you use a GCC version that matches what is used on one of the supported machines (hercules, hera, etc.)

@SamuelTrahanNOAA Thanks for testing the CI fix. The updates to the GCC CI can be submitted in a separate PR. All IFI relevant updates look good to me. Would you like to keep the current CI fix in  your PR for merging, or incorporate it into the CI fix PR?

**Comment by SamuelTrahanNOAA** [comment]
I do not see my changes as a "CI fix." All I did was restore it to the behavior it had before GitHub changed ubuntu-latest to ubuntu 24.04. Any version change of GCC (if you decide to do one) should be in another PR.

**Comment by WenMeng-NOAA** [comment]
@SamuelTrahanNOAA Thanks for clarifying. We will proceed with your PR for the merging process.

**Comment by WenMeng-NOAA** [comment]
@gspetro-NOAA Please go ahead to approve this PR.

**Review by WenMeng-NOAA — APPROVED** [review]

**Comment by WenMeng-NOAA** [comment]
This PR is ready for merging.

**Pull request merged**
